core/src/main/java/tc/oc/pgm/action/replacements/ReplacementParser.java

```diff
@@ -0,0 +1,153 @@
+        throw InvalidXMLException.coerce(e, new Node(el));
+      }
+    } else {
+      throw new InvalidXMLException("Unknown replacement type: " + el.getName(), el);
+    }
+  }
+
+  public Set<String> replacementTypes() {
+    return methodParsers.keySet();
+  }
+
+  protected Method getParserFor(Element el) {
+    return methodParsers.get(el.getName().toLowerCase());
+  }
+
+  private <B extends Filterable<?>> Class<B> parseScope(Element el, Class<B> scope)
+      throws InvalidXMLException {
+    if (scope == null) return Filterables.parse(Node.fromRequiredAttr(el, "scope"));
+    Node node = Node.fromAttr(el, "scope");
+    // Narrower replacements cannot be used with broader scope
+    if (node != null && !Filterables.isAssignable(scope, Filterables.parse(node)))
+      throw new InvalidXMLException(
+          "Wrong scope defined for replacement, scope must be " + scope.getSimpleName()
+              + " or higher",
+          el);
+    return scope;
+  }
+
+  @MethodParser("decimal")
+  public <T extends Filterable<?>> Replacement parseDecimal(Element el, Class<T> scope)
+      throws InvalidXMLException {
+    scope = parseScope(el, scope);
+    var formula = parser.formula(scope, el, "value").required();
+    var format = parser
+        .string(el, "format")
+        .attr()
+        .optional()
+        .<NumberFormat>map(DecimalFormat::new)
+        .orElse(DEFAULT_FORMAT);
+
+    return ScopedReplacement.of(scope, ctx -> text(format.format(formula.applyAsDouble(ctx))));
+  }
+
+  @MethodParser("player")
+  public <T extends Filterable<?>> Replacement parsePlayer(Element el, Class<T> scope)
+      throws InvalidXMLException {
+    var variable = parser.variable(el, "var").scope(MatchPlayer.class).singleExclusive();
+    var fallback = parser.component(el, "fallback").optional(empty());
+    var nameStyle = parser.parseEnum(NameStyle.class, el, "style").optional(NameStyle.VERBOSE);
+    return filterable ->
+        variable.getHolder(filterable).map(mp -> mp.getName(nameStyle)).orElse(fallback);
+  }
+
+  @MethodParser("switch")
+  public <T extends Filterable<?>> Replacement parseSwitch(Element el, Class<T> scope)
+      throws InvalidXMLException {
+    scope = parseScope(el, scope);
+    record SwitchBranch(Component result, Range<Double> valueRange, Filter filter) {}
+
+    var formula = parser.formula(scope, el, "value").orNull();
+    var fallback = parser.component(el, "fallback").child().optional(empty());
+    var children = el.getChildren("case");
+    var branches = new ArrayList<SwitchBranch>(children.size());
+
+    for (var innerEl : children) {
+      var valueRange = formula != null ? parser.doubleRange(innerEl, "match").orNull() : null;
+      var filter = parser.filter(innerEl, "filter").optional(() -> {
+        if (valueRange == null)
+          throw new InvalidXMLException(
+              "At least a filter or a match attribute must be specified", innerEl);
+        return StaticFilter.ALLOW;
+      });
```

Comment:
If you don't specify a value, and do specify a match, you could end up with a weird error message:

"Unused node 'match'" alongside an "At least a filter or a match..." which may leave you puzzled if you were a mapdev

maybe an option would be:
```suggestion
      var filterParse = parser.filter(innerEl, "filter");
      var filter = valueRange == null ? filterParse.required() : filterParse.orAllow();
```
which will just mark filter as required (without mentioning match at all) if you've not specified match.
not sure if it's really better or worse, maybe the err message should be different depending on if value is specified or not? i really don't know what's better

Reply:
> If you don't specify a value, and do specify a match, you could end up with a weird error message:
> 
> "Unused node 'match'" alongside an "At least a filter or a match..." which may leave you puzzled if you were a mapdev

That will never happen. Parsing errors interrupt the whole map loading process, and unused node warnings are the very last step in that process. The error message probably needs a rewrite given that `match="..."` is ignored when there's no value though.